### ISSUE TITLE
fzf-tmux fails with tmux's `run-shell` command

### ISSUE THREAD

**Issue by mike-hearn** [junegunn/fzf#1032]
- Category
    - [ ] fzf binary
    - [x] fzf-tmux script
    - [ ] Key bindings
    - [ ] Completion
    - [ ] Vim
    - [ ] Neovim
    - [ ] Etc.
- OS
    - [x] Linux
    - [x] Docker!
    - [ ] Mac OS X
    - [ ] Windows
    - [ ] Windows Subsystem for Linux
    - [ ] Etc.
- Shell
    - [x] bash
    - [ ] zsh
    - [ ] fish

-----

I'm running into an issue using `fzf-tmux` as part of tmux's `run-shell` command.

If I run `fzf-tmux` from either tmux's built-in command prompt, or inside a tmux session using `tmux run-shell fzf-tmux`, it fails with the following error on 0.17.0 (e89eebb):

```
root@baf17140c5fe:/# tmux run-shell fzf-tmux
'fzf-tmux' returned 2
```

I downgraded to 0.15.9 to try it, and it gives a different (but presumably related) error:

```
root@5f43672d3b38:/# tmux run-shell fzf-tmux
Failed to open /dev/tty
'fzf-tmux' returned 2
```

It's definitely possible I'm missing something super obvious, but I spent some time today and couldn't figure it out.

I made a [Dockerfile](https://github.com/mike-hearn/fzf-tmux-issue/blob/master/Dockerfile) with fzf 0.17.0 and tmux 2.5 to help reproduce the issue. The repo is here with instructions:

https://github.com/mike-hearn/fzf-tmux-issue

Quick screencast of the issue (expected result first, then demo of the error using the Dockerfile):
![Error demo](https://i.imgur.com/ybGWJHp.gif)
[Higher quality MP4 version](https://zippy.gfycat.com/SarcasticFatalBlacknorwegianelkhound.mp4)

**Comment by mike-hearn** [comment]
Update: I just went through my own reproduction steps, using the same Dockerfile, on a remote Linode Debian 8 box, and... it worked as expected.

So there must be some environmental difference between my local Macbook pro, where fzf-tmux fails inside tmux inside Docker (heh), and Debian 8, where fzf-tmux works inside tmux inside Docker. Will keep investigating.

**Comment by mike-hearn** [comment]
A bit more info. Whether this succeeds or fails seems to be based on the underlying operating system, specifically it fails on more recent releases. Notably it works on Fedora 25, but fails on Fedora 26.

I added five additional Dockerfiles to [my example repo](https://github.com/mike-hearn/fzf-tmux-issue); three of them work properly with `fzf-tmux`, while two of them do not.

**Working:**
- [Dockerfile-debian-9-working](https://github.com/mike-hearn/fzf-tmux-issue/blob/master/Dockerfile-debian-9-working)
- [Dockerfile-fedora-22-working](https://github.com/mike-hearn/fzf-tmux-issue/blob/master/Dockerfile-fedora-22-working)
- [Dockerfile-fedora-25-working](https://github.com/mike-hearn/fzf-tmux-issue/blob/master/Dockerfile-fedora-25-working)

**Not working:**
- [Dockerfile-debian-buster-20170723-not-working](https://github.com/mike-hearn/fzf-tmux-issue/blob/master/Dockerfile-debian-buster-20170723-not-working)
- [Dockerfile-fedora-26-not-working](https://github.com/mike-hearn/fzf-tmux-issue/blob/master/Dockerfile-fedora-26-not-working)

For ease of testing, I pushed all these up to Docker Hub.

## Working Case (Fedora 25)

```
$ docker run --rm -ti --hostname=fedora-25 docker.io/mikehearn/fzf-fedora-25 /bin/bash

# Inside container:
[root@fedora-25 /]# tmux

# Inside tmux, inside container:
[root@fedora-25 /]# tmux run-shell fzf-tmux  # works!
```

## Failing Case (Fedora 26)

```
$ docker run --rm -ti --hostname=fedora-26 docker.io/mikehearn/fzf-fedora-26 /bin/bash

# Inside container:
[root@fedora-26 /]# tmux

# Inside tmux, inside container:
[root@fedora-26 /]# tmux run-shell fzf-tmux  # fails!
```

**Comment by mike-hearn** [comment]
Even more info. After chasing down lose ends (originally thought maybe it was `bash` or `tmux` version issues), I finally resorted to multi-stage Dockerfiles to replace files one at a time until `fzf-tmux` worked.

It turned out that Fedora 26 works correctly when replacing the `tput` binary from Fedora 25:

```Dockerfile
FROM fedora:25 as fedora25

FROM fedora:26
RUN dnf install -y tmux git curl
RUN git clone --depth 1 https://github.com/junegunn/fzf.git ~/.fzf && ~/.fzf/install
RUN cp /usr/bin/tput /usr/bin/tput-backup

# THIS IS THE CRUCIAL LINE; WHY DOES IT WORK I DON'T KNOW
COPY --from=fedora25 /usr/bin/tput /usr/bin/tput

# For fingerprinting the different binaries
RUN dnf install -y isomd5sum
```

I have *no idea* why this fixes the issue. They're apparently different, but they return the same `ncurses` version? But their MD5 fingerprints are different. So... they're different. I don't understand.


This is the Fedora 25 version of tput
```
[root@fedora-26-with-tput-fix /]# /usr/bin/tput -V
ncurses 6.0.20170212
[root@fedora-26-with-tput-fix /]# md5sum /usr/bin/tput
c60cce3a95e822660deff375c25ee886  /usr/bin/tput
[root@fedora-26-with-tput-fix /]#
```

This is the Fedora 26 version that we backed up
```
[root@fedora-26-with-tput-fix /]# /usr/bin/tput-backup -V
ncurses 6.0.20170212
[root@fedora-26-with-tput-fix /]# md5sum /usr/bin/tput-backup
6fcf1c30470a7e1763fa059eefb451c1  /usr/bin/tput-backup
[root@fedora-26-with-tput-fix /]#
```

I think this is as far as I can go in my debugging. I don't think it's tenable for Fedora 26 or Debian 10 users to replace their `tput` binaries, so I'm curious if anything can be done on the fzf side to work with the `6fcf1c30470a7e1763fa059eefb451c1` version of tput.

(Also thank you for FZF. I use it a thousand times a day, every day, across bash/vim/tmux. It's amazing.)

**Comment by junegunn** [comment]
Hi, thanks for the investigation. I can reproduce the problem. Using `-x` flag of bash reveals more of what's going on.

```
root@2e7ec047b415:/# tmux run-shell 'bash -x fzf-tmux 2> /tmp/xxx'; cat /tmp/xxx
'bash -x fzf-tmux 2> /tmp/xxx' returned 2
++ command -v fzf
+ fzf=/root/.fzf/bin/fzf
+ [[ -x /root/.fzf/bin/fzf ]]
+ args=()
+ opt=
+ skip=
+ swap=
+ close=
+ term=
+ [[ -n '' ]]
++ tput lines
tput: terminal attributes: No such device or address

+ lines=
+ [[ -n '' ]]
++ tput cols
tput: terminal attributes: No such device or address

+ columns=
+ [[ 0 -gt 0 ]]
+ [[ -z /tmp/tmux-0/default,20,0 ]]
+ [[ '' =~ ^-h ]]
+ [[ ! '' =~ ^-h ]]
+ [[ '' -le 15 ]]
+ /root/.fzf/bin/fzf
Failed to open /dev/tty
+ exit 2
```

- `tput` fails when with `run-shell`
    - `tmux run-shell 'tput lines'`
- Due to the failure of `tput lines` and `tput cols`, fzf-tmux script gets wrong `$lines` and `$columns`, empty strings instead of numbers.
- Which makes it fall back to regular fzf (`[[ "" -lt 40 ]]`), which also fails to read `/dev/tty` and exits with status 2.

Unfortunately, I have no clue why this is happening. Maybe on some OSs, `/dev/tty` is only accessible on interactive context, which actually makes sense.

A workaround would be to assign default values to `$lines` and `$columns` when `tput` fails, but I'm not sure if it's a right thing to do.

**Comment by mike-hearn** [comment]
Looks like tmux has a built-in `pane_height` option that could be helpful. It works where `tput lines` fails in the Fedora 26 container:

```
# All of this is inside tmux:

[root@fedora-26 /]# tput lines
76

[root@fedora-26 /]# tmux run-shell "tput lines"
'tput lines' returned 1

[root@fedora-26 /]# tmux run-shell "tmux display-message -p '#{pane_height}'"
76
```

Just throwing it out there as a possible fallback.

**Comment by mike-hearn** [comment]
Added #1037 to implement this fallback. Not sure if this is all it requires, but it seems to work in my testing on Fedora 26 and Debian 10 (`debian:buster`).

**Comment by junegunn** [comment]
Closed by #1037. Thanks for looking into this.

**Comment by junegunn** [comment]
FYI, `tmux run-shell "tmux display-message -p '#{pane_height}'"` also works outside tmux.
